lib/util/dependent-resource.js

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2017 Balena
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file contains an abstract implementation for dependent metadata resources:
+key-value resources directly attached to a parent (e.g. tags, config variables).
+*/
+
+import * as Promise from 'bluebird';
+
+import {
+	findCallback,
+	isId,
+	isUnauthorizedResponse,
+	mergePineOptions,
+} from '../util';
+import { getUpsertHelper } from '../util/upsert';
+
+export function buildDependentResource(
+	{ pine },
+	{
+		resourceName, // e.g. device_tag
+		resourceKeyField, // e.g. tag_key
+		parentResourceName, // e.g. device
+		getResourceId, // e.g. getId(uuidOrId)
+	},
```

Comment:
This used to include `ResourceNotFoundError` but typescript told me it wasn't used anywhere and made me remove it :P